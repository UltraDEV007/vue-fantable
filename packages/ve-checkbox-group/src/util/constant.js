// prefix
export const PREFIX_CLS = 've-checkbox-group-'

// emit events
export const EMIT_EVENTS = {
  // on checked change
  ON_CHECKED_CHANGE: 'on-checked-change',
}

// comps name
export const COMPS_NAME = {
  VE_CHECKBOX: 'VeCheckbox',
  VE_CHECKBOX_GROUP: 'VeCheckboxGroup',
}
