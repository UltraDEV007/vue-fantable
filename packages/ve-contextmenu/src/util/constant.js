// prefix
export const PREFIX_CLS = 've-contextmenu-'

// comps name
export const COMPS_NAME = {
  VE_CONTEXTMENU: 'VeContextmenu',
}

// emit events
export const EMIT_EVENTS = {
  ON_NODE_CLICK: 'on-node-click',
}

// init data
export const INIT_DATA = {
  // PARENT_ID: 0,
  PARENT_DEEP: -1,
}

// contextmenu types
export const CONTEXTMENU_NODE_TYPES = {
  // Separator
  SEPARATOR: 'SEPARATOR',
}

// instance methods
export const INSTANCE_METHODS = {
  HIDE_CONTEXTMENU: 'hideContextmenu',
}
