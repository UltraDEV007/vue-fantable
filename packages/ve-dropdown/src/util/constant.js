// prefix
export const PREFIX_CLS = 've-dropdown-'

// emit events
export const EMIT_EVENTS = {
  FILTER_RESET: 'on-filter-reset',
  FILTER_CONFIRM: 'on-filter-confirm',
  VISIBLE_CHANGE: 'on-dropdown-visible-change',
  ITEM_SELECT_CHANGE: 'on-item-select-change',
}

// comps name
export const COMPS_NAME = {
  VE_DROPDOWN: 'VeDropdown',
}
