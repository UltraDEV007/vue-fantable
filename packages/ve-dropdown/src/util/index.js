/*
 * @clsName
 * @desc  get class name
 * @param {string} cls - class
 */
export function clsName(cls) {
  return 've-dropdown-' + cls
}
