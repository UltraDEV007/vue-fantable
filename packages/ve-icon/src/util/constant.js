// prefix
export const PREFIX_CLS = 've-icon-'

// emit events
export const EMIT_EVENTS = {}

// comps name
export const COMPS_NAME = {
  VE_ICON: 'VeIcon',
}
