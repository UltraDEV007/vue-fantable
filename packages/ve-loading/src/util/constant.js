// prefix
export const PREFIX_CLS = 've-loading-'

// emit events
export const EMIT_EVENTS = {}

// comps name
export const COMPS_NAME = {
  VE_LOADING: 'VeLoading',
  VE_LOADING_PLANE: 'VeLoadingPlane',
  /* VE_LOADING_CHASE: "VeLoadingChase", */
  VE_LOADING_BOUNCE: 'VeLoadingBounce',
  VE_LOADING_WAVE: 'VeLoadingWave',
  VE_LOADING_PULSE: 'VeLoadingPulse',
  VE_LOADING_FLOW: 'VeLoadingFlow',
  /* VE_LOADING_SWING: "VeLoadingSwing", */
  /*  VE_LOADING_CIRCLE: "VeLoadingCircle",
VE_LOADING_CIRCLE_FADE: "VeLoadingCircleFade", */
  VE_LOADING_GRID: 'VeLoadingGrid',
  VE_LOADING_FOLD: 'VeLoadingFold',
}

// spin name
export const SPIN_NAMES = {
  PLANE: 'plane',
  GRID: 'grid',
  WAVE: 'wave',
  FLOW: 'flow',
  BOUNCE: 'bounce',
  PULSE: 'pulse',
}
