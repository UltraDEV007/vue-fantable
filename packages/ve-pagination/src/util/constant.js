// prefix
export const PREFIX_CLS = 've-pagination-'

// locale comp name
export const LOCALE_COMP_NAME = 'pagination'

// emit events
export const EMIT_EVENTS = {
  PAGE_NUMBER_CHANGE: 'on-page-number-change',
  PAGE_SIZE_CHANGE: 'on-page-size-change',
}

// comps name
export const COMPS_NAME = {
  VE_PAGINATION: 'VePagination',
}
