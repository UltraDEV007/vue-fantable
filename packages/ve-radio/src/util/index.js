import { PREFIX_CLS } from './constant'

/*
 * @clsName
 * @desc  get class name
 * @param {string} cls - class
 */
export function clsName(cls) {
  return PREFIX_CLS + cls
}
