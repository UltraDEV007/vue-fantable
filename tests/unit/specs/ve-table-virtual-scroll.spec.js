import { mount } from '@vue/test-utils'
import FanTable from '@P/fan-table/fan-table'
import { later } from '../util'
import bodyTrScrolling from '@P/fan-table/src/body/body-tr-scrolling.jsx'
import { vi, describe, it, expect, beforeEach, afterEach } from 'vitest'

globalThis.ResizeObserver = class ResizeObserver {
  observe() {
    // do nothing
  }

  unobserve() {
    // do nothing
  }

  disconnect() {
    // do nothing
  }
}
describe('FanTable virtual scroll', () => {
  // same row height
  const TABLE_DATA_SAME_ROW_HEIGHT = []
  for (let i = 0; i < 1000; i++) {
    TABLE_DATA_SAME_ROW_HEIGHT.push({
      rowKey: i,
      name: i,
      date: i,
      hobby: i,
      address: i,
    })
  }

  // different row height
  function getRandom(min, max) {
    return Math.floor(Math.random() * (max - min) + min)
  }

  const TABLE_DATA_DIFFERENT_ROW_HEIGHT = []

  for (let i = 0; i < 1000; i++) {
    let value = i
    if (i % 2 === 0) {
      const rowCount = getRandom(3, 5)

      for (let i = 0; i < rowCount; i++) {
        value += `this is the long word.<br />`
      }
    }

    TABLE_DATA_DIFFERENT_ROW_HEIGHT.push({
      rowKey: i,
      name: value,
      date: i,
      hobby: i,
      address: i,
    })
  }

  //
  const MAX_HEIGHT = 500
  const MIN_ROW_HEIGHT = 40

  // 表格渲染的数量
  // const TABLE_ROW_COUNT = Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) + 1;
  const TABLE_ROW_COUNT = Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) * 2

  // get table rendered row count by row height
  function getTableRenderedRowCountByRowHeight(rowHeight, bufferScale = 2) {
    return Math.ceil(MAX_HEIGHT / rowHeight) * bufferScale
  }

  it('render same row height', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('dynamically turns virtual scrolling on', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: false,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.find('.fan-table-virtual-scroll').exists()).toBe(false)

    wrapper.setProps({
      virtualScrollOption: {
        // 是否开启
        enable: true,
      },
    })

    await later()

    const tableContainerEl = wrapper.find('.fan-table-container')

    expect(tableContainerEl.attributes('style')).toBe(
      'max-height: 500px; height: 500px;',
    )

    expect(wrapper.find('.fan-table-virtual-scroll').exists()).toBe(true)
    expect(wrapper.find('.fan-table-virtual-phantom').exists()).toBe(true)
    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('same row height', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('different row height', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_DIFFERENT_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('buffer scale', async () => {
    const bufferCount = 10

    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_DIFFERENT_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
          bufferScale: bufferCount,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT) * (bufferCount + 1),
    )
  })

  it('with row checkbox', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
            // type=checkbox
            type: 'checkbox',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        checkboxOption: {},
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.find('.ve-checkbox').exists()).toBe(true)

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('with row radio', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
            // type=radio
            type: 'radio',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        radioOption: {},
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.find('.ve-radio').exists()).toBe(true)

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('with row expand', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
            // type=expand
            type: 'expand',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        expandOption: {
          render: ({ row, column, rowIndex }, h) => {
            return (
              <p>
                                My name is{' '}
                <span style="color:#1890ff;">{row.name}</span>
                                ,I'm living in {row.address}
              </p>
            )
          },
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(wrapper.find('.fan-table-row-expand-icon').exists()).toBe(true)

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('with column fixed', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
            fixed: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
            fixed: 'right',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    const thEls = wrapper.findAll(
      '.fan-table-header .fan-table-header-tr .fan-table-header-th',
    )

    expect(thEls[0].classes()).toContain('fan-table-fixed-left')
    expect(thEls[0].classes()).toContain(
      'fan-table-last-left-fixed-column',
    )

    expect(thEls[2].classes()).toContain('fan-table-fixed-right')
    expect(thEls[2].classes()).toContain(
      'fan-table-first-right-fixed-column',
    )

    const tdEls = wrapper.findAll(
      '.fan-table-header .fan-table-header-tr .fan-table-header-th',
    )

    expect(tdEls[0].classes()).toContain('fan-table-fixed-left')
    expect(tdEls[0].classes()).toContain(
      'fan-table-last-left-fixed-column',
    )

    expect(tdEls[2].classes()).toContain('fan-table-fixed-right')
    expect(tdEls[2].classes()).toContain(
      'fan-table-first-right-fixed-column',
    )

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('with column footer', async () => {
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
            fixed: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
            fixed: 'right',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        footerData: [
          {
            rowKey: 0,
            name: '平均值',
            date: 1100,
            hobby: 1200,
            address: 1300,
          },
          {
            rowKey: 1,
            name: '汇总值',
            date: 701000,
            hobby: 801000,
            address: 801000,
          },
        ],
        virtualScrollOption: {
          // 是否开启
          enable: true,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(
      wrapper.findAll('.fan-table-footer-tr .fan-table-footer-td').length,
    ).toBe(6)

    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(
      TABLE_ROW_COUNT,
    )
  })

  it('scrolling callback', async () => {
    const mockFn = vi.fn()
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
          scrolling: mockFn,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    expect(mockFn).toHaveBeenCalledTimes(1)

    expect(mockFn).toHaveBeenCalledWith({
      startRowIndex: 0,
      visibleStartIndex: 0,
      visibleEndIndex: Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT),
      visibleAboveCount: 0,
      visibleBelowCount: Math.ceil(MAX_HEIGHT / MIN_ROW_HEIGHT),
    })
  })

  it('minRowHeight prop', async () => {
    const mockFn = vi.fn()
    const minRowHeight = 50
    const wrapper = mount(FanTable, {
      props: {
        columns: [
          {
            field: 'name',
            key: 'b',
            title: 'Name',
            width: 200,
            align: 'left',
          },
          {
            field: 'hobby',
            key: 'c',
            title: 'Hobby',
            width: 300,
            align: 'left',
          },
          {
            field: 'address',
            key: 'd',
            title: 'Address',
            width: '',
            align: 'left',
          },
        ],
        tableData: TABLE_DATA_SAME_ROW_HEIGHT,
        virtualScrollOption: {
          // 是否开启
          enable: true,
          minRowHeight,
        },
        maxHeight: MAX_HEIGHT,
        rowKeyFieldName: 'rowKey',
      },
    })

    wrapper.triggerResizeObserver({ width: MAX_HEIGHT })

    await later()

    const rowCount = getTableRenderedRowCountByRowHeight(minRowHeight)
    expect(wrapper.findAll('.fan-table-body-tr').length).toBe(rowCount)
  })

  /*
    how to test?
    */
  it('virtual scroll palceholder', async () => {
    const wrapper = mount(bodyTrScrolling, {
      props: {
        colgroups: [{}],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
