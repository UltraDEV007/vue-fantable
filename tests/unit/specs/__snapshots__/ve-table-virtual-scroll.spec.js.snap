// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FanTable virtual scroll render same row height 1`] = `
<div class="vue-table-root">
  <div class="fan-table fan-table-border-around" style="width: 100%;">
    <div class="fan-table-container fan-table-virtual-scroll fan-table-enable-cell-selection" style="max-height: 500px; height: 500px;">
      <div class="fan-table-virtual-phantom fan-table-virtual-scroll" style="height: 40000px;">
        <div style="width: 100%;"></div>
      </div>
      <div class="fan-table-content-wrapper">
        <table class="fan-table-content fan-table-border-x">
          <colgroup>
            <col style="width: 200px;">
            <col style="width: 300px;">
            <col>
          </colgroup>
          <thead class="fan-table-fixed-header fan-table-header">
            <tr class="fan-table-header-tr">
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: left; top: 0px;">Name</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th" style="text-align: left; top: 0px;">Hobby</th>
              <th rowspan="1" colspan="1" class="fan-table-header-th fan-table-last-column" style="text-align: left; top: 0px;">Address</th>
            </tr>
          </thead>
          <tbody class="fan-table-body fan-table-row-hover fan-table-row-highlight">
            <tr style="height: 0px;">
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
              <td style="padding: 0px; border: 0px; height: 0px;"></td>
            </tr>
            <tr row-key="0" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">0</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">0</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">0</td>
            </tr>
            <tr row-key="1" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">1</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">1</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">1</td>
            </tr>
            <tr row-key="2" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">2</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">2</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">2</td>
            </tr>
            <tr row-key="3" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">3</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">3</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">3</td>
            </tr>
            <tr row-key="4" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">4</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">4</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">4</td>
            </tr>
            <tr row-key="5" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">5</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">5</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">5</td>
            </tr>
            <tr row-key="6" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">6</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">6</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">6</td>
            </tr>
            <tr row-key="7" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">7</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">7</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">7</td>
            </tr>
            <tr row-key="8" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">8</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">8</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">8</td>
            </tr>
            <tr row-key="9" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">9</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">9</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">9</td>
            </tr>
            <tr row-key="10" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">10</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">10</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">10</td>
            </tr>
            <tr row-key="11" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">11</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">11</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">11</td>
            </tr>
            <tr row-key="12" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">12</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">12</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">12</td>
            </tr>
            <tr row-key="13" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">13</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">13</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">13</td>
            </tr>
            <tr row-key="14" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">14</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">14</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">14</td>
            </tr>
            <tr row-key="15" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">15</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">15</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">15</td>
            </tr>
            <tr row-key="16" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">16</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">16</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">16</td>
            </tr>
            <tr row-key="17" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">17</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">17</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">17</td>
            </tr>
            <tr row-key="18" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">18</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">18</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">18</td>
            </tr>
            <tr row-key="19" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">19</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">19</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">19</td>
            </tr>
            <tr row-key="20" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">20</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">20</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">20</td>
            </tr>
            <tr row-key="21" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">21</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">21</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">21</td>
            </tr>
            <tr row-key="22" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">22</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">22</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">22</td>
            </tr>
            <tr row-key="23" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">23</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">23</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">23</td>
            </tr>
            <tr row-key="24" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">24</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">24</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">24</td>
            </tr>
            <tr row-key="25" class="fan-table-body-tr">
              <td rowspan="1" colspan="1" col-key="b" class="fan-table-body-td" style="text-align: left;">25</td>
              <td rowspan="1" colspan="1" col-key="c" class="fan-table-body-td" style="text-align: left;">25</td>
              <td rowspan="1" colspan="1" col-key="d" class="fan-table-body-td" style="text-align: left;">25</td>
            </tr>
          </tbody>
          <tfoot class="fan-table-fixed-footer fan-table-footer"></tfoot>
        </table>
        <div class="fan-table-selection-wrapper">
          <div class="fan-table-selection-fixed-left"></div>
          <div class="fan-table-selection-middle"></div>
          <div class="fan-table-selection-fixed-right"></div>
        </div>
      </div>
    </div>
    <div class="fan-table-edit-input-container" style="top: 0px; left: 0px; height: 1px; z-index: -1; opacity: 0;"><textarea tabindex="-1" class="fan-table-edit-input"></textarea></div>
  </div>
</div>
`;

exports[`FanTable virtual scroll virtual scroll palceholder 1`] = `
<tr class="fan-table-body-tr fan-table-body-row-scrolling">
  <td colspan="1"></td>
</tr>
`;
